File: Binary_Search_Trees/733-flood-fill/flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        cur = image[sr][sc]
        if cur == color:
            return image

        n = len(image)
        m = len(image[0])
        q = deque()
        q.append((sr, sc))
        image[sr][sc] = color

        directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        while q:
            row, col = q.popleft()

            for dr, dc in directions:
                newr = dr + row
                newc = dc + col
                if 0 <= newr < n and 0 <= newc < m and image[newr][newc] == cur:
                    q.append((newr, newc))
                    image[newr][newc] = color
        
        return image

File: Binary_Search_Trees/733-flood-fill/flood-fill.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int cur = image[sr][sc];
        if (cur == color) return image;

        int n = image.size();
        int m = image[0].size();
        queue<pair<int, int>> q;
        q.push({sr, sc});
        image[sr][sc] = color;

        vector<pair<int, int>> directions = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

        while (!q.empty()) {
            int row = q.front().first;
            int col = q.front().second;
            q.pop();

            for (auto [dr, dc] : directions) {
                int newr = row + dr;
                int newc = col + dc;

                if (newr >= 0 && newr < n && newc >= 0 && newc < m && image[newr][newc] == cur) {
                    q.push({newr, newc});
                    image[newr][newc] = color;
                }
            }
        }

        return image;
    }
};
